### PR TITLE
Feature/713 result type

### PR DESCRIPTION
**Pull request by mtar**
## Description

This PR provides an implementation of result_type. Additionally, some changes in __binary_op on the output are introduced.

Issue/s resolved: #713 

## Changes proposed:
- add result_type()
- __binary_op calls result_type() for type promotions
- __binary_op uses scalars if an operand is a number
- new entries type mappings

## Type of change
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
yes, the result type of __binary_op follows result_type and scalars are returned when adding scalars and numbers.

### CODE REVIEW

File: heat/core/types.py [hunk 1]
```diff
@@ -854,6 +861,87 @@ def promote_types(
...
+            else:
+                prec1 = 2
+        except Exception:
+            try:
+                # type
+                if isinstance(arg, np.dtype):
+                    arg = arg.char
+                type1 = canonical_heat_type(arg)
+                prec1 = 1
+            except Exception:
```
Comment (coquelin77) [inline heat/core/types.py:907]: why such a broad exception?

File: heat/core/types.py [hunk 2]
```diff
@@ -854,6 +861,87 @@ def promote_types(
...
+
+                tc1 = __type_codes[type1]
+                tc2 = __type_codes[type2]
+
+                if tc1 < tc2:
+                    return type2, min(prec1, prec2)
+                if tc1 > tc2:
+                    return type1, min(prec1, prec2)
+
+                raise RuntimeError("Something went wrong")  # pragma: no cover
```
Comment (coquelin77) [inline heat/core/types.py:937]: can drop these down some tabs, the if loops end with returns

**Review by coquelin77 — CHANGES_REQUESTED** [review]
looks good. minor changes and questions

File: heat/core/types.py (hunk 2 above)
Comment (mtar) [reply heat/core/types.py]: Yes, I flattened the else block.

File: heat/core/types.py (hunk 1 above)
Comment (coquelin77) [reply heat/core/types.py:908]: what is raised when this happens? `TypeError`?

Comment (mtar) [reply heat/core/types.py]: Ah - I first thought you meant the except block and not the caught exception :facepalm:

**Review by coquelin77 — APPROVED** [review]

**Comment by coquelin77** [comment]
will merge once tests finish

**Pull request merged**